sound/soc/codecs/hdac_hda.c

```diff
@@ -519,7 +520,10 @@ static void hdac_hda_codec_remove(struct snd_soc_component *component)
 	pm_runtime_disable(&hdev->dev);
 	snd_hdac_ext_bus_link_put(hdev->bus, hlink);
 
-	snd_hdac_regmap_exit(hdev);
+	if (codec->patch_ops.free)
+		codec->patch_ops.free(codec);
+
+	snd_hda_codec_cleanup_for_unbind(codec);
```

Comment:
wow, this looks like a lot of clean up was missing? I see `snd_hda_codec_cleanup_for_unbind()` is currently called on the clean up path from `hda_codec_driver_remove()`. Is that function not called at all for this codec in the mem-leak case? Should it be called? Just for my information: how are these two clean up routines (codec and component `.remove()`) supposed to work together in this case?

Reply:
@lyakh yeah, maybe I should use stronger wording in the commit message -- we missed releasing 95% of allocated resources. I have looked at the cleanup routines many times, but it never occured to me to check that they are actually called... and it appears they are not. This is not so easy to pick-up with code review. hda_codec_driver_remove() is called but as ASoC/SOF sets the hdac-ext ops, the removal patch just calls "return codec->bus->core.ext_ops->hdev_detach(&codec->core);"  For the latter question, component remove is called when the ASoC card is removed and codec remote is called later when the codec driver module is finally unloaded. The HDA codec remove functions have some bits that assume the parent card to be still around, so we have to do most of the the cleanup in component remove.